Commenting/Implemented like comment use case.

AndreiGanichev opened the pull request:
1. Added `MeetingMemberCommentLike `aggregate to Write model.

1. Added LikesCount to Read model, but didn't add another database table and stored it in existed MeetingComments table are being used for store `MeetingComment` aggregate.

1. Made LikesCount eventually consistent with amount of `MeetingMemberCommentLike` for certain member and comment.

1. Added domain service `MeetingCommentService` to check `CommentCannotBeLikedByTheSameMemberMoreThanOnceRule` because to check the rule out-of-process dependency(database) is required.

@kgrzybek if you agree with the point number 4 than may be we should create base class for domain services. It can contains logic for checking business rules and also allows to add architectures tests for naming.

kgrzybek commented on src/Database/CompanyName.MyMeetings.Database/Scripts/CreateStructure.sql:
```diff
@@ -363,9 +363,24 @@ CREATE TABLE meetings.MeetingComments
+
+CREATE TABLE [meetings].[MeetingMemberCommentLikes]
+(
+	[Id] UNIQUEIDENTIFIER NOT NULL,
+	[MemberId]  UNIQUEIDENTIFIER NOT NULL,
+	[MeetingCommentId] UNIQUEIDENTIFIER NOT NULL,
+    CONSTRAINT [PK_meetings_MeetingMemberCommentLikes_Id] PRIMARY KEY ([Id] ASC),
+    CONSTRAINT [FK_meetings_MeetingMemberCommentLikes_Members] FOREIGN KEY ([MemberId]) REFERENCES meetings.Members ([Id]),
+    CONSTRAINT [FK_meetings_MeetingMemberCommentLikes_MeetingComments] FOREIGN KEY ([MeetingCommentId]) REFERENCES meetings.MeetingComments ([Id])
```
Please fix indents :)

AndreiGanichev replied:
fixed

kgrzybek commented on src/Database/CompanyName.MyMeetings.Database/Scripts/Migrations/1_0_0_0/0013_add_meeting_member_comment_likes_table.sql:
```diff
@@ -0,0 +1,9 @@
+﻿CREATE TABLE [meetings].[MeetingMemberCommentLikes]
+(
+	[Id] UNIQUEIDENTIFIER NOT NULL,
+	[MemberId]  UNIQUEIDENTIFIER NOT NULL,
+	[MeetingCommentId] UNIQUEIDENTIFIER NOT NULL,
+    CONSTRAINT [PK_meetings_MeetingMemberCommentLikes_Id] PRIMARY KEY ([Id] ASC),
+    CONSTRAINT [FK_meetings_MeetingMemberCommentLikes_Members] FOREIGN KEY ([MemberId]) REFERENCES meetings.Members ([Id]),
+    CONSTRAINT [FK_meetings_MeetingMemberCommentLikes_MeetingComments] FOREIGN KEY ([MeetingCommentId]) REFERENCES meetings.MeetingComments ([Id])
```
Please fix indents

AndreiGanichev replied:
fixed

kgrzybek commented on src/Database/CompanyName.MyMeetings.Database/meetings/Tables/MeetingMemberCommentLikes.sql:
```diff
@@ -0,0 +1,10 @@
+﻿CREATE TABLE [meetings].[MeetingMemberCommentLikes]
+(
+	[Id] UNIQUEIDENTIFIER NOT NULL,
+	[MemberId]  UNIQUEIDENTIFIER NOT NULL,
+	[MeetingCommentId] UNIQUEIDENTIFIER NOT NULL,
+    CONSTRAINT [PK_meetings_MeetingMemberCommentLikes_Id] PRIMARY KEY ([Id] ASC),
+    CONSTRAINT [FK_meetings_MeetingMemberCommentLikes_Members] FOREIGN KEY ([MemberId]) REFERENCES meetings.Members ([Id]),
+    CONSTRAINT [FK_meetings_MeetingMemberCommentLikes_MeetingComments] FOREIGN KEY ([MeetingCommentId]) REFERENCES meetings.MeetingComments ([Id])
```
indents

AndreiGanichev replied:
fixed

kgrzybek commented on src/Modules/Meetings/Application/MeetingComments/GetMeetingCommentLikers/GetMeetingCommentLikersQueryHandler.cs:
```diff
@@ -0,0 +1,36 @@
+            var connection = _sqlConnectionFactory.GetOpenConnection();
+
+            var sql = "SELECT " +
+                      "[Liker].[Id]," +
+                      "[Liker].[Name] " +
```
Please use `nameof(MeetingCommentLikerDto.PropertyName)` for each returned property

AndreiGanichev replied:
fixed

kgrzybek commented on src/Modules/Meetings/Domain/MeetingComments/MeetingCommentService.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace CompanyName.MyMeetings.Modules.Meetings.Domain.MeetingComments
+{
+    public class MeetingCommentService
```
This should be in the Application layer, because it is an impure service and it does not represent any business concept. 

Domain Service in Domain should - take some inputs and returns the output - everything based on Domain objects. 

If you want to implement this logic in Domain, you need to set `meetingMemberCommentLikesCount` in Application and pass it as an argument to the method of Domain Object. This way you don't need to mock the repository.

Of course sometimes it is better to keep this logic in application - mostly for performance reasons.

AndreiGanichev replied:
Thanks for your opinion! I get it and will change implementation. As for current implementation I would like do explain my decision:
1. I wanted make Domain model incapsulated(completed). Now it provides to application layer one way to like comment(call `MeetingCommentService` method) and keeps all invariants.
1. I get known about such approach(with impure Domain service) from this article of Vladimir Khorikov. Although he describes the approach he nevertheless tends to keep Domain model free of out-of-process dependencies and make model incomplete but pure. But I decided that inpure Domain service is not so bad if its only dependency is `IMeetingMemberCommentLikesRepository` and as far as I know an interface of repository is considered as part of Domain model.

What do you think about these points? 

And returning to your comment, what do you prefer: place logic in application layer or pass `meetingMemberCommentLikesCount ` as argument?  I think it's better to check the rule in application layer because `meetingMemberCommentLikesCount` as a `AddLike` method's parameter will make `MeetingComment`'s abstraction leaky imo.

kgrzybek replied:
Hi Andrei,

In actual implementation `MeetingCommentService` does not have any business logic in my opinion (only orchestrate application workflow as handlers). I agree with Vladimir that when there is some logic sometimes it is good to extract it to separate service but I would say this is not the case here in my opinion :)

About solution - I always try to put as much as possible logic in the Domain so I would pass it raw data as an argument to the `Like` method. 

Of course, sometimes it is not practical because of performance - then you can pass some abstraction like `ICommentValidator` or make this validation in the application layer.

AndreiGanichev replied:
Ok, thanks for explanation!

kgrzybek commented on src/Modules/Meetings/Application/MeetingComments/AddMeetingCommentLike/AddMeetingCommentLikeCommandHandler.cs:
```diff
@@ -0,0 +1,60 @@
+                throw new InvalidCommandException(new List<string> { "To add like the comment must exist." });
+            }
+
+            var meeting = await _meetingRepository.GetByIdAsync(meetingComment.GetMeetingId());
+
+            var meetingGroup = await _meetingGroupRepository.GetByIdAsync(meeting.GetMeetingGroupId());
+
+            var like = await _meetingCommentService.AddLikeAsync(meetingGroup, meetingComment, _memberContext.MemberId);
+
+            await _meetingMemberCommentLikesRepository.AddAsync(like);
```
This is ok but I would like to point out that in real system it could be a performance problem. You are loading 2 another aggregates (meeting, meetingGroup) only to provide some information (not to protect invariants of these aggregates). Sometimes it is better to consider another approach - to load data via read model and pass to the method.

AndreiGanichev replied:
I also think that load two other aggregates seems to much for such "light" action as liking but i did't find a better implementation. Thanks for explanation about Read model 👍 

>You are loading 2 another aggregates (meeting, meetingGroup) only to provide some information (not to protect invariants of these aggregates)

I need to protect "**Member** can like and unlike the comment" invariant. So MeetingGroup instance was needed to check if a liker is its member. I have looked at IsMemberOfGroup method of MeetingGroup aggregate again and understood that there isn't any logic there(just filter soft deleted members) so it's safe to check the invarinat other way.

AndreiGanichev replied:
Is it ok to create `GetMeetingGroupMemberQuery` with `MemberId` and `MeetingId` values and pass result DTO to `MeetingComment` to check if group member with liker id exists or not?

kgrzybek replied:
Yes it is ok to create separate data class, read it from read model and pass it to the aggregate method.

However, it is not DTO, it is just data, I like to name if with postfix Data and what is important **it is part of the Domain**. Domain can have data classes too - only we need to distinguish them from Entities and ValueObjects.

AndreiGanichev replied:
Interesting insight about models with Data postfix, thanks! Now I've found few and their usages in the project.

>However, it is not DTO, it is just data

DTO's main use case is **transfer** so they should optimized to it(serialization): 
1. primitive types
2. no logic, no encapsulation - just container for data. 

Models with Data postfix are also containers for data but:
1. part of Domain model and so may contain domain objects(mostly ValueObjects, for example TypedIds)
2. aren't supposed to serialize
3. readonly

Am I right about these differences between the concepts?

AndreiGanichev replied:
To continue discussion of concepts and names: what do you think about add ViewModel postfix for QueryHandlers result?

I think it will help to emphisize that the results are supposed to use in Presentation layer. On the contrary there are DTO which are part of Read model but are used in Application layer only.

kgrzybek replied:
I would not call them ViewModel or View because you can have multiple consumers of your module API and API adapter is only one of them. Moreover, API can be consumed by non-GUI process like CLI.

What we return is a just data. I call it with postfix Dto to separate them from other objects. They are transferred between the layers and returned to API consumer. However, they can be called with postfix Data too in my opinion.

AndreiGanichev replied:
Ok

kgrzybek reviewed (changes requested):
I added some things to change and to consider. In general good work 👍

AndreiGanichev commented:
@kgrzybek I removed MeetingCommentService and moved check all business rules into MeetingComment aggregate, replaced loading few aggregate by using Read Model.

Do you have anything to change in the PR now?

kgrzybek commented:
No, looks good now 👍  Thanks for that PR and discussion!
